### ISSUE TITLE
Previewing bookdown chapters in RStudio ignores "knit to" format selection

### ISSUE THREAD

**Issue by connorp** [rstudio/rstudio#5139]
When writing with **bookdown** in RStudio, I often want to preview chapters using `preview_chapter()`. According to the [bookdown documentation](https://bookdown.org/yihui/bookdown/preview-a-chapter.html), `preview_chapter()` should work for all output formats, and elsewhere, [the documentation](https://bookdown.org/yihui/bookdown/rstudio-ide.html) mentions that the **Knit** button in the toolbar should preview the selected chapter. I interpret these two statements together to mean the **Knit** button will preview the selected chapter using `preview_chapter()` in whatever format I select with "Knit to…". I know @yihui et al. recommend only previewing to HTML, but I want to occasionally preview a PDF of a chapter, to review it with some less tech-savvy collaborators (also sometimes I just want to read it on paper). 

This is not the behavior I get from the RStudio IDE **Knit** button. It ignores the selected "Knit to…" format, and calls `render_site()` for the chapter rather than `preview_chapter()`, outputting an HTML file regardless of the selected format. 

### System details

    RStudio Edition : Desktop
    RStudio Version : 1.2.1335
    OS Version      : macOS 10.13.6
    R Version       : 3.5.3

### Steps to reproduce the problem

1. Create a new **bookdown** project (just the minimal project created by 
`File -> New Project… -> Book Project using bookdown` is sufficient).

2. Remove the template bibliography line ~`bibliography: [book.bib, packages.bib]`~ from the YAML header of `index.Rmd` so the project can compile. 

3. Select a particular chapter `.Rmd` file (other than `index.Rmd`) and click **Knit to PDF**. 

### Describe the problem in detail

Instead of knitting a preview PDF of the chapter, RStudio adds the following YAML header to the chapter (which should not be present in a **bookdown** chapter file at all)

```
---
output:
  pdf_document: default
  html_document: default
---
```

And then proceeds to knit an HTML file anyway. The function call is 

```r
rmarkdown::render_site('~/testbook/03-method.Rmd',  encoding = 'UTF-8')
```

without a specification of the output format to `pdf_document2`, `pdf_book`, etc. 

This error also verified in the RStudio Preview version `1.2.1555`, as well as the daily build `1.3.331`, downloaded this morning (July 24, 2019).  (I cannot confirm the function being called by the **Knit** button in the daily build since it is not displayed at the top of the R Markdown console pane as in earlier versions, but an HTML file is still generated when a PDF is requested). 

### What I would expect

When selecting a particular output format via "Knit to…," RStudio should not add a YAML header to a chapter file, and should generate the particular file in the chosen format using `preview_chapter()`

It's possible calling `render_site()` instead of `preview_chapter()` is the intended behavior and I am just misinterpreting the documentation, but I certainly would not expect the selected output format to be ignored, nor RStudio to add a YAML header where it doesn't belong.

**Comment by ronblum** [comment]
@connorp Thank you for raising this. It's understandable why one would expect something like
```
rmarkdown::render('~/testbook/03-method.Rmd',  encoding = 'UTF-8', output_format = "pdf_document")
```
which produces the desired result. We'll review this as we continue development of RStudio.

**Comment by connorp** [comment]
I think for **bookdown** it needs to be `pdf_document2`, but otherwise yes.

**Comment by ronblum** [comment]
Oops, you're right—just because `pdf_document` doesn't mean it does what we want it to do. D'oh.

**Comment by ChrisWaloszek** [comment]
Having the same problem with bookdown – even in a very much reduced MWE:
```
--- 
output: pdf_document2
---
# Random Header
``` 
Trying to compile a bookdown PDF with terminal I get the following error message:
```
 MyMac:BookDownTest User$ bookdown::render_book('index.Rmd', 'bookdown::pdf_book')
-bash: syntax error near unexpected token `'index.Rmd','
```
Thanks for any help!

**Comment by braverock** [comment]
This appears to still be an issue in Current RStudio 1.4.  

I would expect 'knit to PDF' to produce a PDF, not an HTML preview, even when in a bookdown project.

**Comment by cderv** [comment]
@connorp Currently with **bookdown** the knit button will call `render_site()` because a Bookdown project will have a `site: bookdown::bookdown_site` in the YAML header in `index.Rmd`. This is how the RStudio IDE knows this is a site project and that a Rmd document is not on its own. 

This allows the Knit button to call `rmarkdown::render_site` (instead of `rmarkdown::render`) and use the provided site renderer (`bookdown::bookdown_site` in a bookdown project) to render the file. This function in **bookdown** will take care of running the equivalent of `preview_chapter()`. So the calling of `render_site()` is expected here.

However, currently the IDE Knit button does not allow to choose in which format to preview the book. As the call to `render_site()` does not pass any output format, the default one will be used. Default one being the first one you have in the `_output.yaml` file or if no file, in the YAML header in index.Rmd. 
So, as a workaround, you can have set `pdf_book()` as the first format in `_output.yml` in a bookdown project. This would make the Knit button work with the format you want in preview mode (that mean having only the current chapter rendered - other would be empty) Currently, you may have `gitbook()` as the first format so you get HTML. (or not but it is the default for **bookdown**)

This means that indeed the issue is with the Knit button menu that currently does not correctly adapt to a **bookdown** project. It behaves as with a standalone document by adding a YAML header where it should not. 
This button should behave like the Build Book button in the Build pane. (which is use to build the whole book). 

Thanks for pointing this out! 

@braverock that is also what you encounter. The menu below the Knit Button should render in the correct format.

@ChrisWaloszek Your issue is different I think. 
This is R code  
```r
bookdown::render_book('index.Rmd', 'bookdown::pdf_book')
```
So this must be run in R console, not in Terminal as you have done. The error you got is from bash not R, telling you that the command is not correct (you can't have really comma in such command). But it would not work anyway as this is bash. 

Also as a general command `bookdown::pdf_document2` is [a single document format](https://bookdown.org/yihui/bookdown/a-single-document.html) to be used with `rmarkdown::render()`, not with `bookdown::render_book()`

To sum up @ronblum, I believe this is a feature request for the Knit button to be more aware of a **bookdown** project to be able to do the right thing: 
* Retrieving output formats from `index.Rmd` or `_output.yml` as it is done for Render book button
* Pass this format to `rmarkdown::render_site(output_format=)` so that the correct format is rendered by **bookdown**

It is kind of a bug that the Knit button menu still have default field for a Rmd document in a **bookdown** project - usually this document have no YAML except the index.Rmd document.

Happy to discuss it further to think more about what to consider for future versions.

**Comment by yihui** [comment]
> * Pass this format to `rmarkdown::render_site(output_format=)` so that the correct format is rendered by **bookdown**

Also note that this is not specific to **bookdown** but any documents compiled through the Knit button. The output format that users chose is not passed to the render function (or site generator). For an individual Rmd document, when you choose an output format behind the Knit button, what the IDE does is to move the output format to the first in the `output` field in YAML and call `render()` without the specific output format (which means the new first output format will be rendered). It will be great if the IDE doesn't modify the YAML metadata of the user's document, but pass the output format to the render function instead.

**Comment by cderv** [comment]
Oh yes I wasn't complete in the explanation and in my wish: 

The "build book" button allows to select a format and in this case it is passed to `render_site()`. I wish the Knit button would behave similarly when selecting a format ! 😄
